single-post-query-loop-selector.php

```diff
@@ -14,59 +14,103 @@
+	 * applies a filter to adjust the query parameters.
+	 *
+	 * @param array $pre_render The pre-render data for the block.
+	 * @param array $parsed_block The parsed attributes of the block.
+	 *
+	 * @return array The modified pre-render data.
+	 */
+	public function pre_render_block( $pre_render, $parsed_block ) {
+		if (
+			! empty( $parsed_block['attrs']['namespace'] )
+			&& 'creativeandrew/single-post-query-loop-selector' === $parsed_block['attrs']['namespace']
+		) {
+			$this->parsed_block = $parsed_block;
+			add_filter( 'query_loop_block_query_vars', [ $this, 'modify_query_vars' ], 10, 3 );
+		}
 
-add_filter(
-	'pre_render_block',
-	__NAMESPACE__ . '\pre_render_block',
-	10,
-	2
-);
-/**
- * Updates the query on the front end based on custom query attributes.
- *
- * This function is used to modify the query on the front end based on the custom query attributes
- * of a specific block. It checks if the block belongs to a particular namespace and, if so,
- * applies a filter to adjust the query parameters.
- *
- * @param array $pre_render The pre-render data for the block.
- * @param array $parsed_block The parsed attributes of the block.
- *
- * @return array The modified pre-render data.
- */
-function pre_render_block( $pre_render, $parsed_block ) {
-	if ( isset( $parsed_block['attrs']['namespace'] ) && 'creativeandrew/single-post-query-loop-selector' === $parsed_block['attrs']['namespace'] ) {
-		add_filter(
-			'query_loop_block_query_vars',
-			function ( $default_query ) use ( $parsed_block ) {
-				if ( isset( $parsed_block['attrs']['query']['include'] ) ) {
-					$default_query['post__in'] = $parsed_block['attrs']['query']['include'];
-				}
-
-				return $default_query;
-			},
-			10,
-			2
-		);
+		return $pre_render;
+	}
+
+	/**
+	 * Modifies the query vars for the Query Loop block.
+	 *
+	 * This function is used to modify the query vars for the Query Loop block. It checks if the block
+	 * has a custom query attribute and, if so, modifies the query vars accordingly.
+	 *
+	 * @param array    $query Array containing parameters for `WP_Query` as parsed by the block context.
+	 * @param WP_Block $block Block instance.
+	 * @param int      $page  Current query's page.
+	 *
+	 * @return array The modified query vars.
+	 */
+	public function modify_query_vars( $query, $block, $page ) {
+		if ( isset( $this->parsed_block['attrs']['namespace'] ) && 'creativeandrew/single-post-query-loop-selector' === $this->parsed_block['attrs']['namespace'] ) {
 
+			if ( isset( $this->parsed_block['attrs']['query']['include'] ) ) {
+				$query['post__in'] = $this->parsed_block['attrs']['query']['include'];
+			}
+
+			if ( isset( $this->parsed_block['attrs']['query']['postType'] ) ) {
+				$query['post_type'] = $this->parsed_block['attrs']['query']['postType'];
+			}
+		}
+
+		// Our work here is done (backs away slowly).
+		remove_filter( 'query_loop_block_query_vars', [ $this, 'modify_query_vars' ], 10, 3 );
```

Comment:
@jtsternberg 
Thanks for the fixes.

What I don't understand is why we need to remove the filter.

I used Ryan Welcher's example: https://github.com/ryanwelcher/advanced-query-loop/blob/trunk/includes/query-loop.php#L61

Any subsequent query shouldn't be affected if we just add the the namespace check 
`'creativeandrew/single-post-query-loop-selector' === $this->parsed_block['attrs']['namespace']`

But there isn't a reason why to remove the filter per se right?

Reply:
I've had to move on from this project, but I think you're probably right.